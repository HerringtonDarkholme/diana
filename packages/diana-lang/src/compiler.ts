// Compiler for Diana language

import type { ASTNode, ProgramNode, KeyValueNode, ObjectNode, ListNode, StringNode, NumberNode, BooleanNode, IdentifierNode } from './parser'

export function compile(ast: ASTNode): any {
  switch (ast.type) {
    case 'Program': {
      // Merge all top-level KeyValue nodes into an object, skip comments
      const obj: Record<string, any> = {};
      for (const child of (ast as ProgramNode).children) {
        if (child.type === 'KeyValue') {
          const kv = compile(child);
          Object.assign(obj, kv);
        }
        // skip comments
      }
      return obj;
    }
    case 'KeyValue': {
      const { key, value } = ast as KeyValueNode;
      const compiledValue = compile(value);
      let processedKey: string | number | boolean;
      if (key.type === 'KeyPath') {
        // Handle dot notation for nested keys
        if (key.path.length === 1) {
          processedKey = key.path[0];
          return { [processedKey]: compiledValue };
        } else {
          let currentObj: Record<string, any> = {};
          let rootObj = currentObj;
          for (let i = 0; i < key.path.length; i++) {
            const k = key.path[i];
            if (i === key.path.length - 1) {
              currentObj[k] = compiledValue;
            } else {
              currentObj[k] = {};
              currentObj = currentObj[k];
            }
          }
          return rootObj;
        }
      } else if (key.type === 'StringKey') {
        processedKey = key.value;
        return { [processedKey]: compiledValue };
      } else if (key.type === 'ComputedKey') {
        processedKey = typeof key.value === 'boolean' ? String(key.value) : key.value;
        return { [processedKey]: compiledValue };
      } else {
        throw new Error(`Unknown key type: ${(key as any).type}`);
      }
    }
    case 'Object': {
      const obj: Record<string, any> = {};
      for (const prop of (ast as ObjectNode).properties) {
        const kv = compile(prop);
        Object.assign(obj, kv);
      }
      return obj;
    }
    case 'List': {
      return (ast as ListNode).items.map(compile);
    }
    case 'String': {
      return (ast as StringNode).value;
    }
    case 'Number': {
      return (ast as NumberNode).value;
    }
    case 'Boolean': {
      return (ast as BooleanNode).value;
    }
    case 'Null': {
      return null;
    }
    case 'Identifier': {
      // Treat identifiers as strings in output
      return (ast as IdentifierNode).name;
    }
    case 'Comment': {
      // Omit comments from output
      return undefined;
    }
    default:
      throw new Error(`Unknown AST node type: ${(ast as any).type}`);
  }
}
