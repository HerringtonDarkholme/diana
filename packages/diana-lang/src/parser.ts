// Parser for Diana language

import type { Token } from './tokenizer'

export type ASTNode =
  | ProgramNode
  | KeyValueNode
  | ObjectNode
  | ListNode
  | StringNode
  | NumberNode
  | BooleanNode
  | NullNode
  | IdentifierNode
  | CommentNode
  | LetBindingNode

export type KeyNode = 
  | { type: 'KeyPath', path: string[] } // keypath: single or nested.level.key
  | { type: 'StringKey', value: string } // string key: "in quote" or "dot.allowed.in.key"
  | { type: 'ComputedKey', value: number | boolean } // computed key: [123] or [true]/[false]

export interface ProgramNode {
  type: 'Program'
  children: ASTNode[]
}

export interface KeyValueNode {
  type: 'KeyValue'
  key: KeyNode
  value: ASTNode
}

export interface ObjectNode {
  type: 'Object'
  properties: KeyValueNode[]
}

export interface ListNode {
  type: 'List'
  items: ASTNode[]
}

export interface StringNode {
  type: 'String'
  value: string
}

export interface NumberNode {
  type: 'Number'
  value: number
}

export interface BooleanNode {
  type: 'Boolean'
  value: boolean
}

export interface NullNode {
  type: 'Null'
}

export interface IdentifierNode {
  type: 'Identifier'
  name: string
}

export interface CommentNode {
  type: 'Comment'
  value: string
}

export interface LetBindingNode {
  type: 'LetBinding'
  name: string
  value: ASTNode
  annotation?: ASTNode // type annotation, optional
  params?: string[] // for function definitions, optional
}

export function parse(tokens: Token[]): ASTNode {
  let pos = 0

  function peek(n = 0): Token {
    return tokens[pos + n]
  }

  function next(): Token {
    return tokens[pos++]
  }

  function expect(type: string, value?: string) {
    const token = next()
    if (token.type !== type || (value !== undefined && token.value !== value)) {
      throw new Error(`Expected ${type}${value ? `(${value})` : ''} but got ${token.type}${token.value ? `(${token.value})` : ''}`)
    }
    return token
  }

  function skipNewlines() {
    while (peek() && peek().type === 'NEWLINE') next()
  }

  function parseProgram(): ProgramNode {
    const children: ASTNode[] = []
    skipNewlines()
    while (peek() && peek().type !== 'EOF') {
      if (peek().type === 'COMMENT') {
        children.push(parseComment())
        skipNewlines()
      } else if (peek().type === 'LET') {
        children.push(parseLetBinding())
        skipNewlines()
      } else if (peek().type === 'IDENTIFIER') {
        children.push(parseKeyValue())
        skipNewlines()
      } else {
        next() // skip unknown
      }
    }
    return { type: 'Program', children }
  }

  function parseComment(): CommentNode {
    const token = expect('COMMENT')
    return { type: 'Comment', value: token.value || '' }
  }

  function parseKeyValue(): KeyValueNode {
    const keyToken = expect('IDENTIFIER')
    let key: KeyNode
    const raw = keyToken.value || ''
    if (raw.startsWith('[') && raw.endsWith(']')) {
      // Bracketed key: could be string, number, or boolean
      const inner = raw.slice(1, -1)
      if (inner.startsWith('"') && inner.endsWith('"')) {
        // Bracketed string key: ["name.value.test"]
        key = { type: 'StringKey', value: inner.slice(1, -1) }
      } else if (inner === 'true' || inner === 'false') {
        key = { type: 'ComputedKey', value: inner === 'true' }
      } else if (!isNaN(Number(inner))) {
        key = { type: 'ComputedKey', value: Number(inner) }
      } else {
        throw new Error(`Invalid computed key: ${raw}`)
      }
    } else if (raw.startsWith('"') && raw.endsWith('"')) {
      // String key: "in quote" or "dot.allowed.in.key"
      key = { type: 'StringKey', value: raw.slice(1, -1) }
    } else {
      // KeyPath: single or nested.level.key
      key = { type: 'KeyPath', path: raw.split('.') }
    }
    skipNewlines()
    expect('COLON')
    skipNewlines()
    const value = parseValue()
    return { type: 'KeyValue', key, value }
  }

  function parseValue(): ASTNode {
    const token = peek()
    if (!token) throw new Error('Unexpected end of input')
    switch (token.type) {
      case 'STRING':
        next()
        return { type: 'String', value: token.value || '' }
      case 'NUMBER':
        next()
        return { type: 'Number', value: Number(token.value) }
      case 'BOOLEAN':
        next()
        return { type: 'Boolean', value: token.value === 'true' }
      case 'NULL':
        next()
        return { type: 'Null' }
      case 'LBRACE':
        return parseObject()
      case 'LBRACKET':
        return parseList()
      case 'INDENT':
        return parseIndentedList()
      case 'IDENTIFIER':
        // For onelineObj: { a.b.c: 2, c.d: 1 }
        // or for topLevel.nested.value: "of course"
        // treat as string or identifier node
        next()
        return { type: 'Identifier', name: token.value || '' }
      default:
        throw new Error(`Unexpected token: ${token.type}`)
    }
  }

  function parseObject(): ObjectNode {
    expect('LBRACE')
    skipNewlines()
    const properties: KeyValueNode[] = []
    while (peek() && peek().type !== 'RBRACE' && peek().type !== 'EOF') {
      if (peek().type === 'IDENTIFIER') {
        properties.push(parseKeyValue())
        skipNewlines()
        if (peek().type === 'COMMA') {
          next()
          skipNewlines()
        }
      } else {
        next(); // skip unknown
      }
    }
    expect('RBRACE')
    return { type: 'Object', properties }
  }

  function parseList(): ListNode {
    expect('LBRACKET')
    skipNewlines()
    const items: ASTNode[] = []
    while (peek() && peek().type !== 'RBRACKET' && peek().type !== 'EOF') {
      if (peek().type === 'NUMBER' || peek().type === 'STRING' || peek().type === 'BOOLEAN' || peek().type === 'NULL' || peek().type === 'IDENTIFIER') {
        items.push(parseValue())
        skipNewlines()
        if (peek().type === 'COMMA') {
          next()
          skipNewlines()
        }
      } else {
        next(); // skip unknown
      }
    }
    expect('RBRACKET')
    return { type: 'List', items }
  }

  function parseIndentedList(): ListNode {
    expect('INDENT')
    const items: ASTNode[] = []
    
    while (peek() && peek().type !== 'DEDENT' && peek().type !== 'EOF') {
      if (peek().type === 'ASTERISK') {
        next() // consume the *
        skipNewlines()
        
        // Parse the item after the *
        // This could be a simple value or an object with key-value pairs
        const item = parseListItem()
        items.push(item)
        skipNewlines()
      } else {
        next() // skip unknown tokens
      }
    }
    
    if (peek() && peek().type === 'DEDENT') {
      next() // consume DEDENT
    }
    
    return { type: 'List', items }
  }

  function parseListItem(): ASTNode {
    // Check if this is a key-value pair or just a simple value
    const token = peek()
    if (!token) throw new Error('Unexpected end of input in list item')
    
    if (token.type === 'IDENTIFIER' && peek(1) && peek(1).type === 'COLON') {
      // This is a key-value pair, parse as object
      const properties: KeyValueNode[] = []
      
      // Parse the first key-value pair
      properties.push(parseKeyValue())
      skipNewlines()
      
      // Check for additional indented key-value pairs
      while (peek() && peek().type === 'INDENT') {
        next() // consume INDENT
        if (peek() && peek().type === 'IDENTIFIER' && peek(1) && peek(1).type === 'COLON') {
          properties.push(parseKeyValue())
          skipNewlines()
        }
        if (peek() && peek().type === 'DEDENT') {
          next() // consume DEDENT
        }
      }
      
      return { type: 'Object', properties }
    } else {
      // Simple value
      return parseValue()
    }
  }

  function parseLetBinding(): LetBindingNode {
    expect('LET')
    skipNewlines()
    // Parse name
    const nameToken = expect('IDENTIFIER')
    const name = nameToken.value || ''
    skipNewlines()
    let annotation: ASTNode | undefined = undefined
    let params: string[] | undefined = undefined
    // Optional type annotation or function params
    if (peek().type === 'COLON') {
      next() // consume ':'
      skipNewlines()
      // Parse type annotation (as identifier or type expression)
      const typeToken = next()
      annotation = { type: 'Identifier', name: typeToken.value || '' }
      skipNewlines()
    } else if (peek().type === 'LBRACKET' || (peek().type === 'IDENTIFIER' && peek().value === '(')) {
      // Function params (not fully implemented, just collect names for now)
      next()
      while (peek().type !== 'RBRACKET' && !(peek().type === 'IDENTIFIER' && peek().value === ')') && peek().type !== 'EOF') {
        if (peek().type === 'IDENTIFIER' && peek().value !== ')') {
          if (!params) params = []
          params.push(peek().value || '')
        }
        next()
      }
      if (peek().type === 'RBRACKET' || (peek().type === 'IDENTIFIER' && peek().value === ')')) next()
      skipNewlines()
    }
    // Expect '='
    if (peek().type === 'IDENTIFIER' && peek().value === '=') {
      next()
    } else if (peek().type === 'COLON') {
      // fallback, allow 'let x: type = ...'
      next()
    } else {
      expect('IDENTIFIER', '=') // fallback, will error
    }
    skipNewlines()
    // Parse value
    let value: ASTNode
    const nextToken = peek()
    if (nextToken && (nextToken.type === 'STRING' || nextToken.type === 'NUMBER' || nextToken.type === 'BOOLEAN' || nextToken.type === 'NULL' || nextToken.type === 'LBRACE' || nextToken.type === 'LBRACKET' || nextToken.type === 'INDENT' || (nextToken.type === 'IDENTIFIER' && nextToken.value && /^[a-zA-Z_][a-zA-Z0-9_]*$/.test(nextToken.value)))) {
      value = parseValue()
    } else {
      // Collect all tokens until NEWLINE or EOF as a string expression
      let expr = ''
      while (peek() && peek().type !== 'NEWLINE' && peek().type !== 'EOF') {
        expr += (expr ? ' ' : '') + (peek().value || '')
        next()
      }
      value = { type: 'String', value: expr }
    }
    return { type: 'LetBinding', name, value, annotation, params }
  }

  return parseProgram()
}
