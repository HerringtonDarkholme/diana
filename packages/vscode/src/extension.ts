import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
  console.log('Diana configuration language extension is now active!');

  // Register commands
  const helloWorldCommand = vscode.commands.registerCommand('vscode-diana.helloWorld', () => {
    vscode.window.showInformationMessage('Hello from Diana Configuration Language!');
  });

  const restartCommand = vscode.commands.registerCommand('diana.restart', () => {
    vscode.window.showInformationMessage('Diana Language Server restart command - LSP not yet implemented');
  });

  // Register hover provider for Diana files
  const hoverProvider = vscode.languages.registerHoverProvider('diana', {
    provideHover(document, position, _token) {
      const range = document.getWordRangeAtPosition(position);
      const word = document.getText(range);
      
      // Provide basic information about Diana syntax
      if (word === 'true' || word === 'false') {
        return new vscode.Hover('Diana boolean value');
      } else if (word === 'null') {
        return new vscode.Hover('Diana null value');
      } else if (word.match(/^-?\d+$/)) {
        return new vscode.Hover('Diana integer value');
      } else if (word.match(/^-?\d+\.\d+$/)) {
        return new vscode.Hover('Diana float value');
      }

      return new vscode.Hover('Diana configuration language');
    }
  });

  // Register document symbol provider for Diana files
  const documentSymbolProvider = vscode.languages.registerDocumentSymbolProvider('diana', {
    provideDocumentSymbols(document, _token) {
      const symbols: vscode.DocumentSymbol[] = [];
      const text = document.getText();
      const lines = text.split('\n');

      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const trimmed = line.trim();
        
        // Skip comments and empty lines
        if (trimmed.startsWith(';') || trimmed === '') {
          continue;
        }

        // Match key-value pairs
        const keyValueMatch = trimmed.match(/^([^:]+):\s*(.*)$/);
        if (keyValueMatch) {
          const key = keyValueMatch[1].trim();
          const value = keyValueMatch[2].trim();
          
          const range = new vscode.Range(i, 0, i, line.length);
          const selectionRange = new vscode.Range(i, line.indexOf(key), i, line.indexOf(key) + key.length);
          
          let kind = vscode.SymbolKind.Property;
          if (value.startsWith('{') || value === '' && i + 1 < lines.length && lines[i + 1].match(/^  /)) {
            kind = vscode.SymbolKind.Object;
          } else if (value.startsWith('[')) {
            kind = vscode.SymbolKind.Array;
          }

          const symbol = new vscode.DocumentSymbol(
            key,
            value,
            kind,
            range,
            selectionRange
          );
          
          symbols.push(symbol);
        }
      }

      return symbols;
    }
  });

  // Register completion provider for Diana files
  const completionProvider = vscode.languages.registerCompletionItemProvider('diana', {
    provideCompletionItems(_document, _position, _token, _context) {
      const completions: vscode.CompletionItem[] = [];

      // Basic value completions
      const trueCompletion = new vscode.CompletionItem('true', vscode.CompletionItemKind.Value);
      trueCompletion.detail = 'Boolean true value';
      completions.push(trueCompletion);

      const falseCompletion = new vscode.CompletionItem('false', vscode.CompletionItemKind.Value);
      falseCompletion.detail = 'Boolean false value';
      completions.push(falseCompletion);

      const nullCompletion = new vscode.CompletionItem('null', vscode.CompletionItemKind.Value);
      nullCompletion.detail = 'Null value';
      completions.push(nullCompletion);

      return completions;
    }
  });

  // Add subscriptions
  context.subscriptions.push(helloWorldCommand);
  context.subscriptions.push(restartCommand);
  context.subscriptions.push(hoverProvider);
  context.subscriptions.push(documentSymbolProvider);
  context.subscriptions.push(completionProvider);

  // Show activation message
  vscode.window.showInformationMessage('Diana configuration language support activated!');
}

export function deactivate() {
  console.log('Diana extension deactivated');
} 