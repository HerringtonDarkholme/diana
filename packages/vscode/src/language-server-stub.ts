/**
 * Stub implementation for Diana Language Server
 * This is a placeholder for future LSP implementation
 */

import * as vscode from 'vscode';

export class DianaLanguageServerStub {
  private outputChannel: vscode.OutputChannel;

  constructor() {
    this.outputChannel = vscode.window.createOutputChannel('Diana Language Server');
  }

  async start(): Promise<void> {
    this.outputChannel.appendLine('Diana Language Server stub starting...');
    // TODO: Implement actual language server startup
    return Promise.resolve();
  }

  async stop(): Promise<void> {
    this.outputChannel.appendLine('Diana Language Server stub stopping...');
    // TODO: Implement actual language server shutdown
    return Promise.resolve();
  }

  async restart(): Promise<void> {
    await this.stop();
    await this.start();
    this.outputChannel.appendLine('Diana Language Server stub restarted');
  }

  dispose(): void {
    this.outputChannel.dispose();
  }
}

/**
 * Future LSP features to implement:
 * - Syntax validation and error reporting
 * - Auto-completion for keys and values
 * - Hover information
 * - Go to definition
 * - Find all references
 * - Document formatting
 * - Code actions and quick fixes
 * - Semantic highlighting
 * - Document symbols
 * - Workspace symbols
 */ 