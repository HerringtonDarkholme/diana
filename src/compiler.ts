// Compiler for Diana language

import type { ASTNode, ProgramNode, KeyValueNode, ObjectNode, ListNode, StringNode, NumberNode, BooleanNode, IdentifierNode } from './parser'

export function compile(ast: ASTNode): any {
  switch (ast.type) {
    case 'Program': {
      // Merge all top-level KeyValue nodes into an object, skip comments
      const obj: Record<string, any> = {};
      for (const child of (ast as ProgramNode).children) {
        if (child.type === 'KeyValue') {
          const kv = compile(child);
          Object.assign(obj, kv);
        }
        // skip comments
      }
      return obj;
    }
    case 'KeyValue': {
      const { key, value } = ast as KeyValueNode;
      // Handle dot notation for nested objects
      const compiledValue = compile(value);
      const keys = key.split('.');
      let currentObj: Record<string, any> = {};
      let rootObj = currentObj;

      for (let i = 0; i < keys.length; i++) {
        const k = keys[i];
        if (i === keys.length - 1) {
          currentObj[k] = compiledValue;
        } else {
          currentObj[k] = {};
          currentObj = currentObj[k];
        }
      }
      return rootObj;
    }
    case 'Object': {
      const obj: Record<string, any> = {};
      for (const prop of (ast as ObjectNode).properties) {
        const kv = compile(prop);
        Object.assign(obj, kv);
      }
      return obj;
    }
    case 'List': {
      return (ast as ListNode).items.map(compile);
    }
    case 'String': {
      return (ast as StringNode).value;
    }
    case 'Number': {
      return (ast as NumberNode).value;
    }
    case 'Boolean': {
      return (ast as BooleanNode).value;
    }
    case 'Null': {
      return null;
    }
    case 'Identifier': {
      // Treat identifiers as strings in output
      return (ast as IdentifierNode).name;
    }
    case 'Comment': {
      // Omit comments from output
      return undefined;
    }
    default:
      throw new Error(`Unknown AST node type: ${(ast as any).type}`);
  }
}
