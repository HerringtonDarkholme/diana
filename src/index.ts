import fs from 'fs'
import path from 'path'

interface Token<T extends string> {
  type: T
}

interface Indent extends Token<'indent'> {}
interface Deindent extends Token<'deindent'> {}
interface Key extends Token<'key'> {
  path: string[]
}
interface Str extends Token<'str'> {
  delimiter: string
  content: string[]
}
interface Value extends Token<'value'> {
  value: number | string | boolean
}
interface Item extends Token<'item'> {}

type Tokens = Array<
  | Indent
  | Deindent
  | Key
  | Str
  | Value
  | Item
>

function convert(filePath: string) {
  const absolutePath = path.resolve(process.cwd(), filePath)
  const content = fs.readFileSync(absolutePath, 'utf8')
  const tokens = tokenize(content)
  const object = parse(tokens)
  return object
}

const INDENT_SIZE = 2
function tokenize(src: string): Tokens {
  let i = 0
  let tokens: Tokens = []
  let lines = src.split('\n')
  let level = 0
  let inList = false
  while (i < lines.length) {
    let [_, indentation, content = ''] = lines[i].match(/^(\s*)(.+)/) || []
    content = content.replace(/;.*/g, '')
    if (!content) {
      i += 1;
      continue;
    }
    let currentLevel = indentation.length / INDENT_SIZE
    if (currentLevel === level + 1) {
      tokens.push({type: 'indent'})
    } else {
      while (currentLevel < level) {
        if (inList) {
          inList = false
        } else {
          tokens.push({type: 'deindent'})
        }
        level -= 1
      }
    }
    level = currentLevel
    if (content.startsWith('* ')) {
      inList = true
      tokens.push({type: 'item'})
      content = content.slice(2)
      level += 1
    }
    let key = /^([^: ]+)\s*:\s*/.exec(content)
    if (key) {
      tokens.push({
        type: 'key',
        path: key[1].split('.')
      })
      content = content.slice(key[0].length)
    }
    // multiline string
    if (content.startsWith('"""') || content.startsWith("'''")) {
      let delimiter = content.slice(0, 3)
      let strContent = content.slice(3)
      while (!lines[++i].includes(delimiter)) {
        strContent += lines[i]
      }
      let idx = lines[i].indexOf(delimiter)
      strContent += lines[i].slice(0, idx)
      tokens.push({
        type: 'str',
        delimiter,
        content: [strContent],
      })
    }
    if (content) {
      tokens.push({
        type: 'value',
        value: JSON.parse(content.replace(/^'|'$/g, '"')) as any,
      })
    }
    i += 1
  }
  return tokens
}

function parse(tokens: Tokens) {
  if (tokens.length === 0) {
    return null
  }
  const token = tokens[0]
  switch (token.type) {
    case 'key':
      return parseObject(tokens)
    case 'item':
      return parseList(tokens)
    case 'value':
      tokens.shift()
      return token.value
    case 'str':
      tokens.shift()
      return token.content[0]
    default:
      throw new Error('unexpected indent')
  }
}

function parseObject(tokens: Tokens) {
  let obj: Record<string, unknown> = {}
  while (tokens[0]?.type === 'key') {
    let path = tokens[0].path
    let value
    tokens.shift()
    let nextToken = tokens.at(0)
    if (!nextToken) {
      value = null
    } else if (nextToken.type === 'indent') {
      tokens.shift()
      value = parse(tokens)
      if (tokens.at(0)?.type === 'deindent') {
        tokens.shift()
      }
    } else {
      value = parse(tokens)
    }
    deepSet(obj, path, value)
  }
  return obj
}

function deepSet(obj: Record<string, any>, path: string[], value: unknown): unknown {
  if (path.length === 0) {
    return value
  }
  const key = path[0]
  obj[key] = deepSet(obj[key] ?? {}, path.slice(1), value)
  return obj
}

function parseList(tokens: Tokens) {
  let list: unknown[] = []
  while (tokens[0]?.type === 'item') {
    let value
    tokens.shift()
    let nextToken = tokens.at(0)
    if (!nextToken) {
      break
    } else if (nextToken.type === 'indent') {
      tokens.shift()
      value = parse(tokens)
      if (tokens.at(0)?.type === 'deindent') {
        tokens.shift()
      }
    } else {
      value = parse(tokens)
    }
    list.push(value)
  }
  return list
}

console.log(JSON.stringify(convert(process.argv[2]), null, 2))
