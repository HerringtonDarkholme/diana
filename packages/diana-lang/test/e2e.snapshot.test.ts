import { describe, it, expect } from 'vitest';
import { compile } from '../src/compiler';
import { parse } from '../src/parser';
import { tokenize } from '../src/tokenizer';

// Helper function to compile Diana string to JSON
function compileToJson(dianaString: string): any {
  const tokens = tokenize(dianaString);
  const ast = parse(tokens);
  return compile(ast);
}

describe('Diana E2E Snapshot Tests', () => {
  it('should compile simple key-value pairs', () => {
    const input = `
name: "Diana"
version: "0.1.0"
active: true
count: 42
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile nested objects', () => {
    const input = `
database: {
  host: "localhost"
  port: 5432
  credentials: {
    username: "admin"
    password: "secret"
    ssl: true
  }
}
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile arrays and lists', () => {
    const input = `
servers: ["web1", "web2", "web3"]
ports: [80, 443, 8080]
features: ["authentication", "logging", "monitoring"]
mixed: [1, "two", true, null]
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile dot notation keys', () => {
    const input = `
app.name: "MyApp"
app.config.debug: true
app.config.timeout: 30000
server.database.host: "db.example.com"
server.database.port: 5432
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile quoted keys with special characters', () => {
    const input = `
"app.name": "Special Key"
"server-config": "dash-key"
["computed.key"]: "bracket notation"
"spaced key": "with spaces"
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile multi-line strings', () => {
    const input = `
description: """
This is a multi-line string
that spans several lines
and preserves formatting.
"""
sql: """
SELECT * FROM users
WHERE active = true
  AND created_at > '2023-01-01'
ORDER BY created_at DESC;
"""
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile complex nested structures', () => {
    const input = `
; Configuration for a web application
application: {
  name: "E-Commerce Platform"
  version: "2.1.0"
  
  ; Server configuration
  server: {
    host: "0.0.0.0"
    port: 8080
    ssl: {
      enabled: true
      cert: "/path/to/cert.pem"
      key: "/path/to/key.pem"
    }
  }
  
  ; Database settings
  database: {
    type: "postgresql"
    connection: {
      host: "db.internal"
      port: 5432
      database: "ecommerce"
      pool: {
        min: 2
        max: 10
        timeout: 30000
      }
    }
  }
  
  ; Feature flags
  features: {
    payments: true
    recommendations: false
    analytics: true
  }
}

; Environment-specific overrides
environment.production: true
environment.debug: false
logging.level: "info"
logging.outputs: ["console", "file"]

; Service configurations as separate objects
payment_service.name: "payment-gateway"
payment_service.url: "https://api.payments.com"
payment_service.timeout: 5000

email_service.name: "email-service"
email_service.url: "https://api.email.com"
email_service.timeout: 3000
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile mixed data types and null values', () => {
    const input = `
string_value: "hello world"
integer_value: 123
float_value: 45.67
boolean_true: true
boolean_false: false
null_value: null
empty_string: ""
zero_number: 0
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile user configuration using dot notation', () => {
    const input = `
; User 1 configuration
user1.id: 1
user1.name: "Alice"
user1.email: "alice@example.com"
user1.active: true
user1.roles: ["admin", "user"]

; User 2 configuration  
user2.id: 2
user2.name: "Bob"
user2.email: "bob@example.com"
user2.active: false
user2.roles: ["user"]
user2.metadata.last_login: "2023-12-01"
user2.metadata.preferences.theme: "dark"
user2.metadata.preferences.notifications: true

; User list
user_ids: [1, 2]
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile with comments throughout', () => {
    const input = `
; Top-level comment
name: "Test App"  ; inline comment

; Section for API configuration
api: {
  ; Base URL for the API
  base_url: "https://api.example.com"
  
  ; API version
  version: "v1"
  
  ; Timeout settings
  timeouts: {
    connect: 5000  ; Connection timeout in ms
    read: 30000    ; Read timeout in ms
  }
}

; Another top-level setting
debug: true
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });

  it('should compile indented lists with asterisk syntax', () => {
    const input = `
pets:
  * object: 'obj'
    inList: true
  * name: "Fluffy"
    type: "cat"
    age: 3

config:
  * feature: "auth"
    enabled: true
  * feature: "logging"
    enabled: false
    level: "debug"

simple_list:
  * "item1"
  * "item2"
  * "item3"
    `.trim();
    
    const result = compileToJson(input);
    expect(result).toMatchSnapshot();
  });
}); 