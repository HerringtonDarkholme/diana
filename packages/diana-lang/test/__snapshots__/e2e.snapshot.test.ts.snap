// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Diana E2E Snapshot Tests > should compile arrays and lists 1`] = `
{
  "features": [
    "authentication",
    "logging",
    "monitoring",
  ],
  "mixed": [
    1,
    "two",
    true,
    null,
  ],
  "ports": [
    80,
    443,
    8080,
  ],
  "servers": [
    "web1",
    "web2",
    "web3",
  ],
}
`;

exports[`Diana E2E Snapshot Tests > should compile complex nested structures 1`] = `
{
  "application": {
    "database": {
      "connection": {
        "database": "ecommerce",
        "host": "db.internal",
        "pool": {
          "max": 10,
          "min": 2,
          "timeout": 30000,
        },
        "port": 5432,
      },
      "type": "postgresql",
    },
    "features": {
      "analytics": true,
      "payments": true,
      "recommendations": false,
    },
    "name": "E-Commerce Platform",
    "server": {
      "host": "0.0.0.0",
      "port": 8080,
      "ssl": {
        "cert": "/path/to/cert.pem",
        "enabled": true,
        "key": "/path/to/key.pem",
      },
    },
    "version": "2.1.0",
  },
  "email_service": {
    "timeout": 3000,
  },
  "environment": {
    "debug": false,
  },
  "logging": {
    "outputs": [
      "console",
      "file",
    ],
  },
  "payment_service": {
    "timeout": 5000,
  },
}
`;

exports[`Diana E2E Snapshot Tests > should compile dot notation keys 1`] = `
{
  "app": {
    "config": {
      "timeout": 30000,
    },
  },
  "server": {
    "database": {
      "port": 5432,
    },
  },
}
`;

exports[`Diana E2E Snapshot Tests > should compile mixed data types and null values 1`] = `
{
  "boolean_false": false,
  "boolean_true": true,
  "empty_string": "",
  "float_value": 45.67,
  "integer_value": 123,
  "null_value": null,
  "string_value": "hello world",
  "zero_number": 0,
}
`;

exports[`Diana E2E Snapshot Tests > should compile multi-line strings 1`] = `
{
  "description": "
This is a multi-line string
that spans several lines
and preserves formatting.
",
  "sql": "
SELECT * FROM users
WHERE active = true
  AND created_at > '2023-01-01'
ORDER BY created_at DESC;
",
}
`;

exports[`Diana E2E Snapshot Tests > should compile nested objects 1`] = `
{
  "database": {
    "credentials": {
      "password": "secret",
      "ssl": true,
      "username": "admin",
    },
    "host": "localhost",
    "port": 5432,
  },
}
`;

exports[`Diana E2E Snapshot Tests > should compile quoted keys with special characters 1`] = `
{
  "app.name": "Special Key",
  "computed.key": "bracket notation",
  "server-config": "dash-key",
  "spaced key": "with spaces",
}
`;

exports[`Diana E2E Snapshot Tests > should compile simple key-value pairs 1`] = `
{
  "active": true,
  "count": 42,
  "name": "Diana",
  "version": "0.1.0",
}
`;

exports[`Diana E2E Snapshot Tests > should compile user configuration using dot notation 1`] = `
{
  "user1": {
    "roles": [
      "admin",
      "user",
    ],
  },
  "user2": {
    "metadata": {
      "preferences": {
        "notifications": true,
      },
    },
  },
  "user_ids": [
    1,
    2,
  ],
}
`;

exports[`Diana E2E Snapshot Tests > should compile with comments throughout 1`] = `
{
  "api": {
    "base_url": "https://api.example.com",
    "timeouts": {
      "connect": 5000,
      "read": 30000,
    },
    "version": "v1",
  },
  "debug": true,
  "name": "Test App",
}
`;
